builtin/aws/lambda/platform.go

```diff
@@ -0,0 +1,706 @@
+}
+
+const (
+	// The default amount of memory to give to the function invocation, 256MB
+	DefaultMemory = 256
+
+	// How long the function should run before terminating it.
+	DefaultTimeout = 60
+)
+
+const lambdaRolePolicy = `{
+	"Version": "2012-10-17",
+	"Statement": [
+		{
+			"Effect": "Allow",
+			"Principal": {
+				"Service": "lambda.amazonaws.com"
+			},
+			"Action": "sts:AssumeRole"
+		}
+	]
+}`
+
+// SetupRole creates an IAM role that will be used by the Lambda function. The role
+// has the basic lambda execution rules attached to it.
+func (p *Platform) SetupRole(L hclog.Logger, app *component.Source, sess *session.Session) (string, error) {
+	svc := iam.New(sess)
+
+	roleName := "lambda-" + app.App
+
+	L.Info("attempting to retrieve existing role", "role-name", roleName)
+
+	queryInput := &iam.GetRoleInput{
+		RoleName: aws.String(roleName),
+	}
+
+	var roleArn string
+
+	getOut, err := svc.GetRole(queryInput)
+	if err == nil {
+		roleArn = *getOut.Role.Arn
+		L.Info("found existing role", "arn", roleArn)
+		return roleArn, nil
+	}
+
+	L.Info("creating new role")
+
+	input := &iam.CreateRoleInput{
+		AssumeRolePolicyDocument: aws.String(lambdaRolePolicy),
+		Path:                     aws.String("/"),
+		RoleName:                 aws.String(roleName),
+	}
+
+	result, err := svc.CreateRole(input)
+	if err != nil {
+		return "", err
+	}
+
+	roleArn = *result.Role.Arn
+
+	L.Info("created new role", "arn", roleArn)
+
+	aInput := &iam.AttachRolePolicyInput{
+		RoleName:  aws.String(roleName),
+		PolicyArn: aws.String("arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"),
+	}
+
+	_, err = svc.AttachRolePolicy(aInput)
+	if err != nil {
+		return "", err
+	}
+
+	L.Info("attached execution role policy")
+
+	return roleArn, nil
+}
+
+// Deploy deploys an image to AWS Lambda.
+func (p *Platform) Deploy(
+	ctx context.Context,
+	log hclog.Logger,
+	src *component.Source,
+	img *ecr.Image,
+	deployConfig *component.DeploymentConfig,
+	ui terminal.UI,
+) (*Deployment, error) {
+	// Start building our deployment since we use this information
+	deployment := &Deployment{}
+	id, err := component.Id()
+	if err != nil {
+		return nil, err
+	}
+
+	sess, err := utils.GetSession(&utils.SessionConfig{
+		Region: p.config.Region,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	roleArn := p.config.RoleArn
+
+	if roleArn == "" {
+		arn, err := p.SetupRole(log, src, sess)
+		if err != nil {
+			return nil, err
+		}
+
+		roleArn = arn
+	}
+
+	mem := int64(p.config.Memory)
+	if mem == 0 {
+		mem = DefaultMemory
+	}
+
+	timeout := int64(p.config.Timeout)
+	if timeout == 0 {
+		timeout = DefaultTimeout
+	}
+
+	lamSvc := lambda.New(sess)
+	curFunc, err := lamSvc.GetFunction(&lambda.GetFunctionInput{
+		FunctionName: aws.String(src.App),
+	})
+
+	var funcarn string
+
+	// If the function exists, we update it's code rather than create a new one.
+	// Will then publish this new code to create a stable identifier for it.
+	if err == nil {
```

Comment:
Should we specifically check for a type of `ResourceNotFoundException` on this err, and return the `err` if it's of another type?

Reply:
I realize this is confusing placement -- this comment should really go on L265, around the `else` clause.

Reply:
Cleaned that up.